### DISCUSSION TITLE
private docker image with docker-compose issue

### DISCUSSION THREAD

**Discussion by s1rd4v3** [Dokploy/dokploy#123] [Q&A] [answered]
Hi all,

I'm having a hard time deploying my nextjs, strapi & postgress application on dockploy with docker-compose. I have an organization on github with a private monorepo. On branch push I've setup a workflow to build and publish both docker images, the strapi app and the nextjs app, against the github registry. This all works fine and as expected and I can directly use my docker-compose.yml with a reference to the images on a Portainer instance without issues. But with Dokploy I'm getting this error about being unauthorized to access a manifest.... No idea what this means to be honest since I haven't heard of those manifest in a docker context yet :). I've authorized Dokploy to access my private repo via the dedicated Settings page (Github App), in the Cluster settings I've added the github registry with my credentials, and I've additionally authenticated via Terminal on my VPS directly via `docker login ghrc.io`. I can successfully pull those two images with `docker pull` independently but just not via docker compose.

Any ideas what I'm missing?

```bash
╔══════════════════════════════════════════════════════════════════════════════╗
║                                                                              ║
║ App Name: compose-connect-1080p-matrix-991r83                                ║
║ Build Compose 🐳                                                             ║
║ Detected: 0 mounts 📂                                                        ║
║ Command: docker compose -p compose-connect-1080p-matrix-991r83 -f            ║
║ ./docker-compose.yml up -d --build --remove-orphans                          ║
║ Source Type: docker github ✅                                                ║
║ Compose Type: docker-compose ✅                                              ║
║                                                                              ║
╚══════════════════════════════════════════════════════════════════════════════╝
 backend Pulling 
 frontend Pulling 
 frontend Error Head "https://ghcr.io/v2/ORGNAME/REPONAME/manifests/frontend": unauthorized
 backend Error context canceled
Error response from daemon: Head "https://ghcr.io/v2/ORGNAME/REPONAME/manifests/frontend": unauthorized
ERROR: Error: child exited with code 18: ❌
```

**Comment by Siumauricio** [comment]
Got it, The registry is used purely on docker swarm, not for docker compose, This is a good case where we can handle this, Because I didn't take in consideration this way like to use a private image, let's move to an issue since this is really a need!

**Comment by Siumauricio** [comment, marked as answer]
I created a issue #125

**Comment by Siumauricio** [comment]
@s1rd4v3 I was a long time looking for a solution to this but apparently you can not mount the authentication file from the container to the host, the way to solve it is to pass this flag with the path where the authentication json is, I made some small changes in the code that when you use the `TEST REGISTRY` button that is in the part where you add the registry, when you press that button and the connection is correct you will get a toaster, that means that the authentication was added correctly to the file /root/. docker/config.json, so now to be able to use that file you would have to assign this environment variable in the `Enviroment Variables` tab like this `DOCKER_CONFIG=/root/.docker/config.json`.

***Note***, In the new update v0.2.0 just make sure to recreate the registry and use the `TEST REGISTRY`  please and that will be enough and also don't forget to add the env `DOCKER_CONFIG=/root/.docker/config.json`.

https://github.com/containrrr/watchtower/discussions/596

Before using the enviroment, error authenticated
![Screenshot 2024-06-07 at 12 18 29 AM](https://github.com/Dokploy/dokploy/assets/47042324/74f261e7-9eda-4cbb-b218-6a19e1557754)

After settting the env
![Screenshot 2024-06-07 at 12 19 01 AM](https://github.com/Dokploy/dokploy/assets/47042324/b379f664-8a16-48b7-8f5f-7c8ebce867f0)

The image is being downloaded and executed sucessfully!
![Screenshot 2024-06-07 at 12 19 12 AM](https://github.com/Dokploy/dokploy/assets/47042324/cdc14e54-31aa-47eb-b95b-42debcc2b2da)

Let me know if you have some questions or bugs!

**Reply by s1rd4v3** [reply]
Thanks a lot for your quick investigations @Siumauricio!!

From a UX perspective, isn't it easier for your users, if the docker config file gets referenced automatically?
Or like having an explicit checkbox on the General -> Providers -> Github config container to append this env key value `DOCKER_CONFIG=/root/.docker/config.json` automatically by dokploy if the user want's to allow authentication with the added cluster registry credentials (If defined).

**Reply by Siumauricio** [reply]
Good point, I'll add the variable by default, but can override by specifying the same variable in the env tab

**Reply by s1rd4v3** [reply]
Sounds great 🙏

**Reply by Siumauricio** [reply]
I Release the new version, let me know if its working!

**Reply by s1rd4v3** [reply]
It's working now. Thanks a lot @Siumauricio 🎉

**Reply by Siumauricio** [reply]
I'm glad that is working!!

**Comment by abdussamadbello** [comment]
@Siumauricio 
Can we do the same thing for dokcer deployments?
<img width="541" alt="image" src="https://github.com/user-attachments/assets/7fe9fcb0-3621-4048-91be-7fd820dfddfa">

**Comment by ivandzen** [comment]
Does the issue still exists in **Version v0.28.8**? I can not run my docker-compose service with `pull_policy: always` Always got error
<img width="976" height="395" alt="image" src="https://github.com/user-attachments/assets/7c72e901-9bff-4a21-a7ed-eb67244aef7e" />

But i already set up credentials in Registry section (Test Registry works) and also logged in in docker itself (using one VM as both dockploy and worker)

**Reply by Alain-Bearez** [reply]
After many attempts, I managed to get that working with **v0.30.2** considering some error messages I got in the mean time.

In your compose, you need to refer to your image by its complete path: `registry.host/registry/image:version`, for me it was `container-registry.br-ne1.magalu.cloud/cuali/app:0.9.2`, where the authorization token is stored in the `DOCKER_CONFIG` file.

Its seems like the image referenced by @ivandzen is missing the registry host part.
